### ISSUE TITLE
File dialogs don't show up

### ISSUE THREAD

**Issue by jpoprof** [flathub/org.octave.Octave#26]
Hi,

I just have downloaded the flatpak with octave 4.4.1 and I noticed that for me all file related dialogues don't appear on screen. The code returns immediately without any error.
I run the flatpak with Xfce on a Debian machine (in a VM).
In parallel there is a "normal octave installed with version 4.0.3.

How to reproduce: 
`>flatpak run org.octave.Octave --gui&`

`>>uigetfile('*.csv')`
`ans = 0`
Code returns immediately. No GUI is shown.
`>>edit junk`
`>>`
Here the popup box for confirming that the file schould be generated appears, but no file dialog.

The same code on Windows with Octave 4.4.0 works as expected.
When I close the file dialog for "edit junk" on Windows without selecting a file then an error message apears in the octave output. This is not the case with those "non-existing" dialogs in the flatpak 4.4.1.

Thanks for the flatpak, since the debian packets for octave are out of date...

Jens

**Comment by mtmiller** [comment]
These examples all work for me on a Debian Gnome desktop, I am not able to reproduce.

> Thanks for the flatpak, since the debian packets for octave are out of date...

The latest version of Octave for Debian is [here](https://packages.debian.org/experimental/octave).

**Comment by Alexander-Wilms** [comment]
Can't reproduce this either on Fedora KDE. Are there .m-files that shadow the proper functions? Does it work with the experimental Debian packages?

**Comment by Alexander-Wilms** [comment]
Does the regular file dialog (File > Open) work?

**Comment by kupiqu** [comment]
I cannot reproduce this either in KDE neon on top of Ubuntu 18.04. However, I frequently see the following warning message:

```
trying to load "/usr/lib/plugins/kf5/kio/file.so" from "/usr/lib/plugins/kf5/kio/file.so"
```

This is a minor thing, but is there a way to silence it?

Sorry if this is not the place to mention this, but it is related to file dialogs.

**Comment by Alexander-Wilms** [comment]
Maybe it's helpful to compare the different implementations.

Here's the C++ code that gets called when executing uigetfile.m: http://hg.savannah.gnu.org/hgweb/octave/file/2ee4d384224f/libgui/graphics/__init_qt__.cc#l167

Will look for the code that's called when opening a file via the main menu.

**Comment by kupiqu** [comment]
But please note issue #28 

So, that log message cannot be seen in the regular dialog as it doesn't use native KDE dialogs at the moment, right?

**Comment by mtmiller** [comment]
Is this only related to file dialogs? Can you try another command that influences the GUI? For example, hide the Documentation panel, then type `doc` in the command window. Does the Documentation panel appear and get focused?

And the `edit` command doesn't normally show a file dialog.

**Comment by jpoprof** [comment]
Closing the doc -panel and then opening it with 'doc' works as expected.
Also opening a message box with msgbox('Hello') works fine.
I just tested the command 'uigetdir()' and it does show a file dialog but also some error messages.

`Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.`
`(octave-gui:4): dconf-WARNING **: failed to commit changes to dconf: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown`

Opening a new figure window with the command 'figure' works also.

So, in total it seems that just the uigetfile command is affected and uigetdir is somewhat shaky (but is usable)

**Comment by kupiqu** [comment]
might it be that org.kde.Sdk/x86_64/5.11 runtime is somehowmissing? If so I think just typing in a terminal: `flatpak update` should fix it.

(if not, maybe the one is missing is the gtk/gnome one if you are in a gnome system and it uses differently).

**Comment by jpoprof** [comment]
I have just performed a `flatpak update` and it updated something.
But the behaviour of octave stayed the same as before.

**Comment by mtmiller** [comment]
This problem may be related to the choice of native vs Qt dialog boxes. @jpoprof can you try a couple more test cases and let us know if the dialog boxes appear or not?

1. Choose "Find Files" from the "Edit" menu. A dialog will show up. The second item in the first column is labeled "Start in". Click on the corresponding "Browse" button. Does a file dialog show up called "Set Search Directory"?

2. Open the "Preferences" dialog from the "Edit" menu. At the bottom of the first tab labeled "General" is a "Browse" button to choose a directory to start Octave in. Does that dialog show correctly?

The dialogs showed by uigetdir, uigetfile, and these two options in the GUI are all native file dialog boxes, as opposed to every other file dialog box shown by Octave. Maybe that will help someone find if there is any knowledge on the web about problems with Qt native dialog boxes in flatpak running under Xfce.

**Comment by jpoprof** [comment]
I did test the following:

1.  Find Files from Edit: the directory browser is shown and works (the selected directory changes the "Start in" text field. But the same error messages appear on the console by the dozens as when I run  'uigetdir()' .
1. "Preferences" dialog from the "Edit": the directory browser is shown and works (the selected directory changes the corresponding text field. But the same error messages appear on the console as when I run  'uigetdir()' .

**Comment by mtmiller** [comment]
Ok, so native file dialogs to pick files are not working at all, and native file dialogs to pick directories are working but spewing diagnostic messages to stderr.

Does `uigetfile` with no arguments work?

If it does, then it may be something about native file dialogs in your environment that have a file extension filter applied.

It it does, then it seems like any native file dialog for selecting files does not work in your environment.

We don't know why that is, we need more information about whether it's your operating system (does another OS VM work?) your desktop environment (does another window manager / desktop session work?), your version of flatpak (can you try installing flatpak from stable-backports?), or some other configuration particular to your system?

**Comment by jpoprof** [comment]
next test:
I installed a brand new Debian 9.5 (downloaded today, debian-9.5.0-amd64-netinst.iso) with the net-installer in a virtualbox VM.
I have chosen Gnome, Xfce and MATE (with the latter apparently chosen as active) as GUI.
During Install I have chosen the following options (see install screenshot on the right):

![grafik](https://user-images.githubusercontent.com/26304207/45368279-81854a80-b5e3-11e8-969b-989ba13893f6.png)

I was later asked to specify one of the concurrent displaymanagers and have chosen lightdm.

When I then install the flatpak and let it run I get a slightly different behaviour regarding the file dialogs:
uigetdir(): works as intended but generates lots of error messages.
uigetfile(): hangs the interpreter: The GUI still reacts (menus are working as well as the editor tab etc. but I am unable to continue entering commands at the command line and I am unable to close the Octave window (I tried the "X" on the window, menu/file/quit ctrl+Q).
I can (while the command window still hangs) edit a new file and save it to disk with a file dialog.
In order to close the flatpak-octave with the hanging interpreter I have to "kill -9" the octave-gui-process.

I later installed the latest octave version from the debian stretch stable which is 4.0.3.3.
With that instance everything works as expected.

**Comment by mtmiller** [comment]
I just tried to replicate your environment. I started with a freshly installed minimal Debian 9.5 VM.

1. I first installed Openbox (`apt install openbox`), a minimal window manager. I then installed Octave from flathub. In an Openbox session running in a VNC, I ran Octave. Both the `uigetdir` and `uigetfile` commands returned immediately with a return value of zero. No dialog box was shown at all in either case. So I was able to reproduce that this condition is possible.

2. I next installed Xfce (`apt install task-xfce-desktop`). I then started an Xfce desktop session in a VNC and ran Octave again. This time, both the `uigetdir` and `uigetfile` commands work correctly. Both commands show a dialog box and behave appropriately.

3. I then went back to the Openbox window manager, but with all of the packages installed by `task-xfce-desktop` still installed. This time, both the `uigetdir` and `uigetfile` commands still work correctly.

So I can indeed replicate this problem when I have a minimal or partially installed system, but not when there is a full Xfce desktop environment properly installed. So technically somewhat reproducible, but resolved for me by installing more packages from the native OS to get a fully functioning desktop environment.

It seems to me that you may be missing some packages or configuration on your system, but I do not know enough about the inner workings of flatpak to even guess where to look.

I will note that if I uninstall the Debian packages xdg-desktop-portal and xdg-desktop-portal-gtk, the `uigetdir` and `uigetfile` commands go back to returning immediately and not showing a dialog box. Do you have these packages installed? Do you see the corresponding processes running in your Xfce session (`ps x | grep xdg`)?

Does anyone else here have any suggestions for what pieces of flatpak desktop integration may be missing here?

**Comment by jpoprof** [comment]
Both packages were already installed.
`root@debian9:/var/log/installer# apt install xdg-desktop-portal
Paketlisten werden gelesen... Fertig
Abhängigkeitsbaum wird aufgebaut.       
Statusinformationen werden eingelesen.... Fertig
xdg-desktop-portal ist schon die neueste Version (0.5-1).
xdg-desktop-portal wurde als manuell installiert festgelegt.
0 aktualisiert, 0 neu installiert, 0 zu entfernen und 1 nicht aktualisiert.
`
`
root@debian9:/var/log/installer# apt install xdg-desktop-portal-gtk
Paketlisten werden gelesen... Fertig
Abhängigkeitsbaum wird aufgebaut.       
Statusinformationen werden eingelesen.... Fertig
xdg-desktop-portal-gtk ist schon die neueste Version (0.5-1).
xdg-desktop-portal-gtk wurde als manuell installiert festgelegt.
0 aktualisiert, 0 neu installiert, 0 zu entfernen und 1 nicht aktualisiert.
`
root@debian9:/var/log/installer# ps aux |grep xdg
lpo+  2638  0.0  0.8 592388  8884 ?        Ssl  08:51   0:00 /usr/lib/flatpak/xdg-document-portal
jpo+  2641  0.0  0.7 368260  7712 ?        Ssl  08:51   0:00 /usr/lib/flatpak/xdg-permission-store
root      4855  0.0  0.0  12784   976 pts/0    S+   10:39   0:00 grep xdg

-------------------------
I was just beginning to read the documentation of flatpak and this came to my attention:
[docs.flatpak.org/en/latest/sandbox-permissions.html:](http://docs.flatpak.org/en/latest/sandbox-permissions.html)

**dconf access**

Until a sandbox-compatible backend is available, applications that require access to dconf can do so with the following options:

--filesystem=xdg-run/dconf
--filesystem=~/.config/dconf:ro
--talk-name=ca.desrt.dconf
--env=DCONF_USER_CONFIG_DIR=.config/dconf

----------------
I was just wondering that none of the example dconf settings show up in the yaml file for octave (org.octave.Octave.yaml) and that the uigetdir()-error messages mention **dbus**. So, there might be a link in between.

**Comment by Alexander-Wilms** [comment]
You can add these permissions after `flatpak run`. That was you can test it without rebuilding the Flatpak.

**Comment by mtmiller** [comment]
From your comment above:

    root@debian9:/var/log/installer# ps aux |grep xdg
    lpo+ 2638 0.0 0.8 592388 8884 ?    Ssl 08:51 0:00 /usr/lib/flatpak/xdg-document-portal
    jpo+ 2641 0.0 0.7 368260 7712 ?    Ssl 08:51 0:00 /usr/lib/flatpak/xdg-permission-store
    root 4855 0.0 0.0 12784 976 pts/0  S+  10:39 0:00 grep xdg

I have a few questions about this.

1. Was there a copy/paste error or do you have two different user accounts, one starting with "lpo" and one starting with "jpo"?

2. Are you testing all of this with the same user account? A normal non-root user is logging in to a graphical session from lightdm and running octave with flatpak, not using su or sudo?

3. As you can see, you don't have any `xdg-desktop-portal` processes shown above. Was this just after logging in, or was this after trying to call `uigetfile` in octave? Please try launching octave in flatpak, running `uigetdir` and `uigetfile`, and then show the output of `ps x | grep xdg`.

**Comment by jpoprof** [comment]
I tried the following:
`flatpak run --filesystem=xdg-run/dconf --filesystem=~/.config/dconf --talk-name=ca.desrt.dconf --env=DCONF_USER_CONFIG_DIR=.config/dconf org.octave.Octave --gui&`

Unfortunately the behaviour stayed the same: Same dconf-warning messages from uigetdir() and the same hanging interpreter for uigetfile().

I then rebooted the VM.
Right after login `ps aux |grep xdg` gave this result:
`jpo+  1393  0.0  0.0  12784   924 pts/0    R+   18:24   0:00 grep xdg`

After starting the flatpak with the command shown above:
```
jpo+  1431  0.0  1.1 592388 11332 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-document-portal
jpo+  1434  0.0  0.9 368260  9744 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-permission-store
jpo+  1472  0.0  0.0  12784   976 pts/0    S+   18:26   0:00 grep xdg
```

After first use of uigetdir() **(Note: without any dconf warning messages!)**
```
jpo+  1431  0.0  1.1 592388 11332 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-document-portal
jpo+  1434  0.0  0.9 368260  9744 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-permission-store
jpo+  1577  0.0  0.0  12784   960 pts/0    R+   18:28   0:00 grep xdg
```

After first use of uigetfile() **(Note: it worked this time!)**
```
jpo+  1431  0.0  1.1 592388 11528 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-document-portal
jpo+  1434  0.0  1.1 375172 11604 ?        Ssl  18:25   0:00 /usr/lib/flatpak/xdg-permission-store
jpo+  1579  0.0  0.9 384532  9348 ?        Sl   18:28   0:00 /usr/lib/xdg-desktop-portal/xdg-desktop-portal
jpo+  1585  2.0  3.3 677624 34304 ?        Sl   18:28   0:02 /usr/lib/xdg-desktop-portal/xdg-desktop-portal-gtk
jpo+  1605  0.0  0.0  12784   944 pts/0    S+   18:31   0:00 grep xdg
```

### It works! It works! It works!

So, to sum it up: it seems that the culprit is the missing permission to access dconf files.
It is possible to set the permissions at runtime. But setting them in the yaml file would be better!
Strange that a reboot was needed to make those permissions on the command line effective.

Note, that I modified the example of the flatpak documentation in that I didn't include the `:ro` suffix.

**Comment by mtmiller** [comment]
Is this the first time that you restarted the VM? Specifically, did you `apt install flatpak` and then `flatpak install` in the same session that you tried to run Octave in without any restarting at all?

Now that you have the app working, can you try to `flatpak uninstall` and reinstall it without changing the dconf permissions and see if it still works?

**Comment by mtmiller** [comment]
I just tested yet again in a brand new VM and I am unable t reproduce this issue at all, so I still do not see that any changes are necessary.

I started with a fresh install of a minimal Debian 9.5 x86_64 VM. I created a user account, installed a full Xfce environment, and started a desktop session under VNC. I then installed flatpak and octave, just to see if the order of installation or rebooting matters, and the `uigetdir` and `uigetfile` commands still work correctly.

I do get the following diagnostic output in the command window

```
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.

(octave-gui:4): dconf-WARNING **: failed to commit changes to dconf: GDBus.Error:org.freede
sktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown
```

when calling `uigetdir`, but it behaves correctly. There is no such output in the command window with `uigetfile`, and it works completely as expected. Same with `uiputfile`.

**Comment by Alexander-Wilms** [comment]
Did the updates to Qt 5.12 and Octave 5 make a difference?

**Comment by Alexander-Wilms** [comment]
Did the updates to Qt 5.12 and Octave 5 make a difference?

**Comment by mtmiller** [comment]
I was never able to reliably reproduce this, other than by installing a minimal system that was missing some components or permissions.

Also note that in Octave 5 the user can toggle a setting between native and non-native file dialogs, so if the default doesn't work for one particular user's system they can easily switch to non-native file dialogs.

**Comment by Alexander-Wilms** [comment]
Should we close this?

**Comment by mtmiller** [comment]
Closing, not reproducible.

**Comment by vryy** [comment]
I installed Octave on flatpak on Ubuntu 18.04. Running it as sudo is fine but when running with user mode, this error below occurred and the GUI does not show:

Gtk-Message: 16:35:31.359: Failed to load module "canberra-gtk-module"

**Comment by cbm755** [comment]
I get the Gtk warning too, I think its harmless.  Octave 5.1  doesn't show the gui unless you explicitly ask for it:
```
Practice problems from "APEX Calculus":Apex12.5: 7, 11, 13, 17, 19, 21, 23, 25, 27.Practice problems from 
```
flatpak run org.octave.Octave --gui
```

**Comment by vryy** [comment]
@cbm755 it's exactly what I ran and had that error

![Octave Critical Error_003](https://user-images.githubusercontent.com/13781578/58065630-8fe32780-7bb0-11e9-94fe-c9623e58588d.png)

**Comment by cbm755** [comment]
But that is not the error you showed before...!

This bug report is closed, and this not the place for support.
(but I would guess from the error messages that you have some permissions problem in your home directory, e.g., files owned by root).

**Comment by vryy** [comment]
@cbm755 you're right. I fixed this.
